test/properties.js

```diff
@@ -0,0 +1,237 @@
+/* eslint-env mocha */
+/* eslint-disable rapid7/static-magic-numbers */
+'use strict';
+
+global.Log = new (require('winston').Logger)();
+
+const Properties = require('../lib/properties');
+const Source = require('../lib/source/common');
+
+// Shorten build hold-down timeout for testing
+Properties.BUILD_HOLD_DOWN = 100;
+
+const expect = require('chai').expect;
+
+class Parser {
+  constructor() {
+    this.properties = {};
+  }
+
+  update(data) {
+    this.properties = data;
+  }
+}
+
+class Stub extends Source(Parser) { // eslint-disable-line new-cap
+  constructor(opts, properties) {
+    // Inject defaults into options
+    const options = Object.assign({
+      type: 'stub',
+      delay: 250 + Math.floor(Math.random() * 250),
+      nopoll: true
+    }, opts);
+
+    super(options);
+    this.delay = options.delay;
+    this.nopoll = options.nopoll;
+    this.properties = properties || {};
+  }
+
+  initialize() {
+    super.initialize();
+
+    // Kill the polling interval.
+    if (this.nopoll) clearTimeout(this._timer);
+  }
+
+  _fetch(callback) {
+    // Simulate a network request
+    setTimeout(() => {
+      callback(null, this.properties);
+    }, this.delay);
+  }
+}
+
+class NoStub extends Stub {
+  _fetch(callback) {
+    callback(null, Source.NO_EXIST);
+  }
+}
+
+describe('Properties', function _() {
+  const properties = new Properties();
+
+  it('adds static properties and builds', function __(done) {
+    properties.static({
+      hello: 'world'
+    });
+
+    expect(properties.layers.length).to.equal(1);
+
+    properties.once('build', (props) => {
+      expect(props).to.be.instanceOf(Object);
+      expect(props.hello).to.equal('world');
+
+      done();
+    });
+
+    properties.build();
+  });
+
+  it('adds layers with namespaces', function __(done) {
+    properties.static({
+      cruel: 'world'
+    }, 'goodbye');
+
+    properties.once('build', (props) => {
+      expect(props.hello).to.equal('world');
+      expect(props.goodbye.cruel).to.equal('world');
+      done();
+    });
+
+    properties.build();
+  });
+
+  it('adds a dynamic layer and rebuilds on updates', function __(done) {
+    const stub = new Stub({}, {
+      stubby: 'property!'
+    });
+
+    properties.dynamic(stub);
+
+    properties.once('build', (props) => {
```

Comment:
Need to initialize the properties instance to register for the layer's update event